Add connection specification to mirror creation

josephsnyder opened the pull request:
This allows each mirror to contain information about the credentials
used to access it.

josephsnyder committed:
Add connection specification to mirror creation

This allows each mirror to contain information about the credentials
used to access it.

Update command and tests based on comments

Switch to only "long form" flags for the s3 connection information.
Use the "any" function instead of checking for an empty list when looking
for s3 connection information.

Split test to use the access token separately from the access id and key.
Use long flag form in test.

Add endpoint_url to available S3 options.

Extend the special parameters for an S3 mirror to accept the
endpoint_url parameter.

Add a test.

josephsnyder commented:
@scottwittenburg, here is my second pull request!

opadron commented on lib/spack/spack/cmd/mirror.py:
```diff
@@ -93,7 +92,15 @@ def setup_parser(subparser):
         default=spack.config.default_modify_scope(),
         help="configuration scope to modify")
-
+    add_parser.add_argument(
+        '-s3-id', '--s3-access-key-id',
+        help="ID string to use to connect to this S3 mirror")
+    add_parser.add_argument(
+        '-s3-key', '--s3-access-key-secret',
+        help="Secret string to use to connect to this S3 mirror")
+    add_parser.add_argument(
+        '-s3-token', '--s3-access-token',
+        help="Access Token to use to connect to this S3 mirror")
```
I'm not a fan of these short flags that aren't actually short.  Better would be to just have the long options, imo.

opadron commented on lib/spack/spack/mirror.py:
```diff
@@ -463,7 +483,15 @@ def add(name, url, scope):
-    items.insert(0, (name, url))
+    mirror_data = url
+    key_values = ["s3_access_key_id", "s3_access_token"]
+    if [value for value in key_values if value in args]:
```
```suggestion
    if any(value for value in key_values if value in args):
```

opadron commented on lib/spack/spack/util/s3.py:
```diff
@@ -31,8 +31,16 @@ def create_s3_session(url):
+    session = Session(**s3_connection)
     s3_client_args = {"use_ssl": spack.config.get('config:verify_ssl')}
 
     endpoint_url = os.environ.get('S3_ENDPOINT_URL')
```
I don't mean to derail the PR, as it's looking pretty good otherwise, but I recall that the endpoint url was another relevant piece of connection information, and it seems like this PR would be an obvious place to include it, along with the keys and session token.

For context, this is an environment variable that you can set to access s3-compliant endpoints other than AWS.

Would be nice to see support for this field added, but this is really an FYI, more than anything else.

opadron reviewed (dismissed):
Marking this review as "approved" though I would like to see the nonstandard cli options changed or removed.

Good stuff! :+1:

josephsnyder committed:
Add connection information per URL not per mirror

Expand the mirror-based connection information to be per-URL.
This will allow a user to specify different S3 connection information
for both the fetch and the push URLs.

Add a parameter for "profile", another way of storing the id/secret pair.

josephsnyder commented:
@opadron, I've finally reached a "happy" point with this now url-specific connection information.  PTAL!

opadron commented on lib/spack/spack/cmd/common/arguments.py:
```diff
@@ -328,3 +328,22 @@ def reuse():
+        '--s3-access-token',
+        help="Access Token to use to connect to this S3 mirror")
+    subparser.add_argument(
+        '--s3-access-profile',
```
This should probably just be `--s3-profile`.

opadron commented on lib/spack/spack/cmd/mirror.py:
```diff
@@ -149,7 +148,15 @@ def mirror_set_url(args):
 
     entry = mirrors[args.name]
-
+    key_values = ["s3_access_key_id", "s3_access_token", "s3_access_profile"]
```
"s3_profile"

opadron commented on lib/spack/spack/cmd/mirror.py:
```diff
@@ -149,7 +148,15 @@ def mirror_set_url(args):
+                         "access_pair": (args.s3_access_key_id,
+                                         args.s3_access_key_secret),
+                         "access_token": args.s3_access_token,
+                         "access_profile": args.s3_access_profile,
```
"profile: args.s3_profile

opadron commented on lib/spack/spack/mirror.py:
```diff
@@ -137,24 +142,83 @@ def __repr__(self):
+            if url_type == "push":
+                return self._push_url['access_pair']
+            return self._fetch_url['access_pair']
+        else:
+            return None
+
+    def set_access_pair(self, url_type, connection_tuple):
+        if url_type == "push":
+            self._push_url["access_pair"] = connection_tuple
+        else:
+            self._fetch_url["access_pair"] = connection_tuple
+
+    def get_endpoint_url(self, url_type):
+        if isinstance(self._fetch_url, dict):
+            if url_type == "push":
+                return self._push_url['endpoint_url']
+            return self._fetch_url['endpoint_url']
+        else:
+            return None
+
+    def set_endpoint_url(self, url_type, url):
+        if url_type == "push":
+            self._push_url["endpoint_url"] = url
+        else:
+            self._fetch_url["endpoint_url"] = url
+
+    def get_access_token(self, url_type):
+        if isinstance(self._fetch_url, dict):
+            if url_type == "push":
+                return self._push_url['access_token']
+            return self._fetch_url['access_token']
+        else:
+            return None
+
+    def set_access_token(self, url_type, connection_token):
+        if url_type == "push":
+            self._push_url["access_token"] = connection_token
+        else:
+            self._fetch_url["access_token"] = connection_token
+
```
Let's make these properties like the others.

josephsnyder replied:
@opadron, the properties and the setters do not accept arguments.  Is it the best plan to have 8 properties, one push and one fetch each, or 4 functions that take an argument like above?

opadron replied:
Oh wait, yes of course, you're right!  I didn't notice the extra arguments.  I retract this feedback. :)

opadron commented on lib/spack/spack/mirror.py:
```diff
@@ -463,7 +527,18 @@ def add(name, url, scope):
     items = [(n, u) for n, u in mirrors.items()]
-    items.insert(0, (name, url))
+    mirror_data = url
+    key_values = ["s3_access_key_id", "s3_access_token", "s3_access_profile"]
```
"s3_profile"

josephsnyder committed:
Switch from "access_profile" to "profile"

alecbcs commented:
Hey @opadron and @josephsnyder looks like this PR broke Autamus's implementation/utilization of build caches and now for we're getting `TypeErrors` every build like [this](https://github.com/autamus/registry/runs/4287857436?check_suite_focus=true). Did this PR unexpectedly change the format for how mirrors are defined? Autamus defines the build cache as an entry in the default `spack.yaml` environment [here](https://github.com/autamus/registry/blob/main/default.yaml) and then we pipe in secrets for writing the cache back through docker's secrets system as ENV variables. Any help to fix this would be greatly appreciated so we can get back to building containers as soon as possible for the community.

josephsnyder commented:
Hi @alecbcs,  Thanks for the comment and the report .  It does change certain aspects of the mirror objects targeted to your use case of S3 URLs. 

I'll try to walk through the steps that you linked and see if I run into issues.  Hopefully we can get you up and running soon!

josephsnyder commented:
@alecbcs, I'm sorry to say that I'm not able to replicate this problem locally, using the E4S stack and a local MinIO instance.  Do you have a good way to get a `spack -d` into the above command so that we can see the debug output?  Or, a pointer to how I can get these GitHub tasks running locally?  

Thanks in advance!

alecbcs commented:
@josephsnyder I'm able to reproduce this error locally with the `spack.yaml` linked above from Autamus. If you have access to a docker install here are the commands that I ran.

```bash
docker run --rm -it spack/ubuntu-bionic
```

This'll drop you into an interactive docker session where I ran,
```bash
apt-get update
apt-get install nano

nano spack.yaml

spack -e . install
```

Here is the exact spack.yaml file I used from Autamus.
```yaml
spack:
  specs:
  - zlib
  view: true
  packages:
    all:
      target: [x86_64_v3]
  config:
    concretizer: clingo
    compiler:
      target: [x86_64_v3]
    install_missing_compilers: true
    install_tree:
      root: /opt/software
      padded_length: 512
  container:
    strip: true
    os_packages:
      build:
      - python3-boto3
      - python3-dev
  mirrors:
    autamus: s3://autamus-cache
```

which resulted in,

```
root@1b55f268eb24:~# spack -e . install
==> Installing environment /root
==> Installing zlib-1.2.11-wlypcgtia47krfm7srhhimn5ijpzfjjo
==> Error: Failed to install zlib due to TypeError: string indices must be integers
==> Error: string indices must be integers
root@1b55f268eb24:~# 
```

Thanks for taking a look at this and let me know if there's anything else I can do to help!

josephsnyder commented:
Thank you @alecbcs.  Your instructions certainly got me there faster.  I can replicate the problem and I'll hopefully have a solution for you soon.

josephsnyder commented:
Okay! @alecbcs  and @opadron, there are a few things that will need an update.  

The update that I am least sure about is the fact that Mirror's `to_dict` function returns a string when no push URL is set: https://github.com/spack/spack/blob/develop/lib/spack/spack/mirror.py#L93.  That is the direct cause of the error above.  I'm not sure what effect changing this to return a dictionary in all code paths will have on the rest of the SPACK code, so I will explore this further and see what further changes are necessary.  

Additionally, once past that, there are a few issues with the accessing of the data in the mirror information, if that mirror information isn't present at all: https://github.com/spack/spack/blob/develop/lib/spack/spack/util/s3.py#L46-L61. There needs to be a consistent way of checking for presence  of the string in the `connection` object and that the value is useful.  These checks will be errors with your current set up because you don't provide the YAML configuration for them.  This is my fault because these values are likely not needed all at once.

The quickest, and dirtiest, way to get you moving again would be to change the mirror object your `spack.yaml` to avoid these pitfalls until they are fixed.  

```
mirrors:
  autamus:
    fetch:
      url: s3://autamus-cache
      access_pair:
      - null
      - null
      access_token: null
      profile: null
    push:
      url: s3://autamus-cache
      access_pair:
      - null
      - null
      access_token: null
      profile: null
```

The script above plus this mirror update gets my script past the TypeError that you reported.  It still fails with a GPG issue, but I think that's a bit outside of this current issue.

opadron commented:
@josephsnyder For `Mirror.to_dict()`, I think changing it to return `dict` in all code paths should be fine.  Any call sites (of which there should be few) would have had to handle both `dict` and non-`dict`, anyway.  I'd be really surprised if such a change actually broke anything.

alecbcs commented:
@josephsnyder would it be possible to have

```yaml
  mirrors:
    autamus: s3://autamus-cache
```
resolve automatically to,
```yaml
mirrors:
  autamus:
    fetch:
      url: s3://autamus-cache
      access_pair:
      - null
      - null
      access_token: null
      profile: null
    push:
      url: s3://autamus-cache
      access_pair:
      - null
      - null
      access_token: null
      profile: null
```

so that we don't continue to break backwards compatibility? 

cc'ing @vsoch since she's also been working on some updates to the Spack mirror code.

vsoch commented:
I did get most of the way through a refactor, but most of the tests were broken so I gave up there and then just did a dummy example (https://github.com/autamus/spack-build-cache-client) of my vision for (hopefully) someone else to tackle, and someone that has experience with the hairball of code that is mirrors!

vsoch commented:
And here is the same integrated into spack https://github.com/spack/spack/compare/develop...vsoch:add/github-packages-build-cache if anyone wants to take a shot at working on the tests - would so appreciate!

josephsnyder commented:
@alecbcs,

The code should already do that, if you go through the `spack mirror add ...` command.  As for your use case,  the updates in #27694 makes it so that the value given for the mirror no longer assumes that there will be a dictionary of information.

stephenmsachs commented:
I am running into issues when adding an existing spack mirror in s3. Also the susequent calls to import keys fail.
I am running the following:

```
[root@ip-172-31-7-191 spack]# spack_mirror="infospherespackmirrortest"
[root@ip-172-31-7-191 spack]# spack mirror add --scope site ${spack_mirror} s3://${spack_mirror}
[root@ip-172-31-7-191 spack]# spack mirror list
infospherespackmirrortest    CommentedMap([('url', 's3://infospherespackmirrortest'), ('access_pair', [None, None]), ('access_token', None), ('profile', None), ('endpoint_url', None)]) (fetch)
infospherespackmirrortest    CommentedMap([('url', 's3://infospherespackmirrortest'), ('access_pair', [None, None]), ('access_token', None), ('profile', None), ('endpoint_url', None)]) (push)
spack-public                 https://mirror.spack.io
[root@ip-172-31-7-191 spack]# spack buildcache keys -i -t
==> Error: sequence item 1: expected str instance, NoneType found
[root@ip-172-31-7-191 spack]# git checkout HEAD~1
Previous HEAD position was d759612523 Add connection specification to mirror creation (#24244)
HEAD is now at a5c50220db py-werkzeug: add 2.0.2 (#27555)
[root@ip-172-31-7-191 spack]# rm etc/spack/mirrors.yaml 
rm: remove regular file ‘etc/spack/mirrors.yaml’? y
[root@ip-172-31-7-191 spack]# spack mirror add --scope site ${spack_mirror} s3://${spack_mirror}
[root@ip-172-31-7-191 spack]# spack mirror list
infospherespackmirrortest    s3://infospherespackmirrortest
spack-public                 https://mirror.spack.io
[root@ip-172-31-7-191 spack]# spack buildcache keys -i -t
gpg: key CAED3547: public key "infospherespackmirrortest (GPG created for Spack) <ec2-infosphere@amazon.com>" imported
gpg: Total number processed: 1
gpg:               imported: 1  (RSA: 1)
gpg: inserting ownertrust of 6
[root@ip-172-31-7-191 spack]# 
```

josephsnyder commented:
Hello @stephenmsachs, I believe the linked pull request (#27694) should get you past this issue.  I ran your sequence on that branch and was able to get to the point where the s3 bucket returned an access error.  Checking out the commit before it causes the error that you linked above.

We're trying to get it merged to develop as we speak.

ashermancinelli commented:
I am also experiencing an issue with an existing buildcache in an s3 bucket. After using the mirror config schema above in a `spack ci generate` gitlab pipeline, the child pipeline fails due to invalid syntax in the config.

Here's the change I made to the environment:
```diff
   mirrors:
-    pnnl: s3://spack-ci
+    pnnl:
+      fetch:
+        url: s3://spack-ci
+        access_pair:
+        - null
+        - null
+        access_token: null
+        profile: null
+      push:
+        url: s3://spack-ci
+        access_pair:
+        - null
+        - null
+        access_token: null
+        profile: null
```

With the short gitlab CI error:
```
variables config should be a hash of key value pairs, value can be a hash
```
[And here's the generated pipeline.](https://github.com/spack/spack/files/7688286/pipeline.4.yml.txt) Potentially related to #27865.

CC @scottwittenburg, as we've been discussing this offline.

stephenmsachs commented:
> Hello @stephenmsachs, I believe the linked pull request (#27694) should get you past this issue. I ran your sequence on that branch and was able to get to the point where the s3 bucket returned an access error. Checking out the commit before it causes the error that you linked above.

Hi @josephsnyder. Thanks for the pointer. I added comment to that PR.

haampie commented on lib/spack/spack/util/s3.py:
```diff
@@ -11,14 +11,23 @@
 import spack.util.url as url_util
 
 
+def get_mirror_connection(url, url_type="push"):
```
@josephsnyder what was the thinking here? Why push_url? It's also used in `get_object` requests? Is this a bug?
